File: auto_video_create_front/src/app/utils/blogUrl.ts
/**
 * 블로그 주소 정규화 + 검증 (2026-08-27).
 *
 * ## 왜 필요한가
 *
 * 신규 유저 ssonek 의 첫 세션(07:49~07:58) 로그에서 두 가지가 나왔다.
 *
 * 1. **빈 주소로 4번 제출됐다** (`요청 블로그: /`). 입력 검증이 아예 없어서
 *    빈 칸으로도 제출이 되고, 유저는 왕복 한 번을 버린 뒤에야 실패를 본다.
 * 2. **같은 주소가 두 번 붙어서 404** 가 났다.
 *    `https://blog.naver.com/ssonek/224391607813https://blog.naver.com/ssonek/224391607813`
 *    붙여넣기가 안 먹은 줄 알고 한 번 더 붙이면 이렇게 된다. 더 나쁜 건 이게
 *    서버의 블로그 검증을 **통과한다**는 점이다 — 호스트가 blog.naver.com 이고
 *    경로 첫 조각이 본인 아이디라 통과한 뒤, 실제로 긁으러 가서 404 가 난다.
 *    유저에게는 이유를 알 수 없는 실패로만 보인다.
 *
 * ## 설계 메모
 *
 * - 순수 함수로 두어 서버 왕복 없이 즉시 피드백을 준다.
 * - 지원하지 않는 호스트는 여기서 걸러도 **결과가 달라지지 않는다**. 서버의
 *   parse_blog_url 이 이미 그런 주소를 빈 username 으로 처리해 거부하기 때문이다.
 *   다만 메시지가 "등록된 블로그 주소가 아닙니다" 라 원인을 알 수 없어서,
 *   여기서 먼저 무엇이 잘못됐는지 알려준다.
 * - 서버 쪽 방어(호스트 부분일치 우회, 구형 PostView 링크)는 별건으로 남아 있다.
 *   이 파일은 유저가 실제로 겪은 두 경우만 다룬다.
 */

export type BlogUrlResult =
  | { ok: true; url: string }
  | { ok: false; message: string };

/** 서버 parse_blog_url 이 실제로 해석할 수 있는 호스트만 통과시킨다. */
function isSupportedHost(host: string): boolean {
  return (
    host === "blog.naver.com" ||
    host === "m.blog.naver.com" ||
    host === "tistory.com" ||
    host.endsWith(".tistory.com") ||
    host === "brunch.co.kr" ||
    host === "m.brunch.co.kr"
  );
}

/**
 * 입력을 정규화하고 검증한다. 성공하면 서버로 보낼 주소를 돌려준다.
 *
 * 정규화 순서가 중요하다 — 앞뒤 공백 제거 → 중복 붙여넣기 잘라내기 →
 * 스킴 보정 → 파싱 → 호스트 확인. 중복 잘라내기를 파싱보다 먼저 해야
 * 두 번 붙은 주소가 "유효한 URL" 로 파싱돼 통과하는 걸 막을 수 있다.
 */
export function normalizeBlogUrl(raw: string): BlogUrlResult {
  let s = (raw ?? "").trim();

  if (!s) {
    return { ok: false, message: "블로그 주소를 입력해주세요." };
  }

  // 주소가 여러 번 들어있으면 첫 번째만 쓴다.
  // 앞에 제목 같은 텍스트가 붙어 오는 경우(네이버 앱 공유)도 같이 걸러진다.
  const starts: number[] = [];
  const scheme = /https?:\/\//gi;
  let m: RegExpExecArray | null;
  while ((m = scheme.exec(s)) !== null) {
    starts.push(m.index);
  }
  if (starts.length > 1) {
    s = s.slice(starts[0], starts[1]).trim();
  } else if (starts.length === 1) {
    s = s.slice(starts[0]).trim();
  } else {
    // 스킴이 없으면 붙여준다. blog.naver.com/foo 처럼 입력하는 경우가 흔하다.
    s = `https://${s}`;
  }

  // 잘라낸 뒤 끝에 남을 수 있는 구분자 정리
  s = s.replace(/[\s,]+$/, "");

  let parsed: URL;
  try {
    parsed = new URL(s);
  } catch {
    return {
      ok: false,
      message: "주소 형식이 올바르지 않아요. 블로그 글 주소를 그대로 붙여넣어 주세요.",
    };
  }

  if (parsed.protocol !== "http:" && parsed.protocol !== "https:") {
    return {
      ok: false,
      message: "주소 형식이 올바르지 않아요. 블로그 글 주소를 그대로 붙여넣어 주세요.",
    };
  }

  const host = parsed.hostname.toLowerCase();
  if (!isSupportedHost(host)) {
    return {
      ok: false,
      message: "네이버 블로그, 티스토리, 브런치 주소만 사용할 수 있어요.",
    };
  }

  if (!isPostUrl(host, parsed)) {
    return {
      ok: false,
      message: "블로그 홈 주소예요. 만들고 싶은 글을 연 뒤 그 글의 주소를 복사해서 넣어주세요.",
    };
  }

  return { ok: true, url: parsed.toString() };
}

/**
 * 개별 **글** 주소인가, 블로그 홈인가 (2026-08-30).
 *
 * 호스트만 맞으면 통과시키던 탓에 `m.blog.naver.com/아이디`(블로그 홈)가 그대로
 * 서버까지 갔다. 크롤러가 본문을 못 찾아 실패하고, 그게 장애 알람까지 울렸다.
 * 유저에게는 "다른 글로 시도해주세요" 라고만 보여서 **무엇이 잘못됐는지 알 수 없었다.**
 *
 * 여기서 걸러 왕복을 없애고, 무엇을 해야 하는지 바로 알려준다.
 */
function isPostUrl(host: string, parsed: URL): boolean {
  const parts = parsed.pathname.split("/").filter(Boolean);

  if (host === "blog.naver.com" || host === "m.blog.naver.com") {
    // 구형 공유 링크: /PostView.naver?blogId=..&logNo=..
    if (parsed.searchParams.get("logNo")) return true;
    // 신형: /{아이디}/{글번호}
    return parts.length >= 2 && /^\d+$/.test(parts[1]);
  }

  if (host === "brunch.co.kr" || host === "m.brunch.co.kr") {
    return parts.length >= 2;             // /@작가/글번호
  }

  // 티스토리 — /{글번호} 또는 /entry/{제목}
  return parts.length >= 1 && parts[0] !== "category";
}

File: auto_video_create_server/services/blog_url.py
"""블로그 주소 해석 — 플랫폼/작성자/글 식별 (2026-08-30).

## 왜 생겼나

두 가지 사고가 같은 뿌리였다.

**하나.** test 알람 (2026-08-30 12:51). 유저가 `m.blog.naver.com/auctionrun0643`
(블로그 **홈**)을 넣었다. 크롤러가 본문을 못 찾아 ValueError 를 던졌고, catch-all 이
이걸 "예상치 못한 실패"로 보고 알람을 울렸다. 하지만 이건 장애가 아니라 **주소를
잘못 넣은 것**이다. 이런 알람이 쌓이면 진짜 장애를 무시하게 된다.

그렇다고 이 에러를 통째로 조용히 만들면 안 된다 — 네이버가 마크업을 바꿔 파서가
깨져도 같은 ValueError 가 나기 때문이다(코드에 그 경고가 남아 있다). **글 주소인데도
본문을 못 찾은 경우와, 애초에 글 주소가 아니었던 경우를 갈라야** 한다.

**둘.** api/blog.py 의 기존 parse_blog_url 에는 알려진 결함이 둘 있었다.

    "blog.naver.com" in host          # 부분 문자열 검사
    → blog.naver.com.attacker.io/유저아이디 가 통과한다.
      "본인 블로그만" 이라는 규칙이 무력화된다.

    path_parts[0] 만 보고 작성자를 뽑음
    → blog.naver.com/PostView.naver?blogId=X&logNo=Y (네이버가 실제로 주는 형식)
      에서 작성자를 'PostView.naver' 로 읽어 본인 글인데도 차단된다.

두 문제 다 "주소를 제대로 해석한다"는 한 가지 일이라 여기로 모은다.
"""
from __future__ import annotations

from typing import NamedTuple, Optional
from urllib.parse import parse_qs, urlparse

NAVER_HOSTS = {"blog.naver.com", "m.blog.naver.com"}
BRUNCH_HOSTS = {"brunch.co.kr", "m.brunch.co.kr"}

# 네이버가 글 목록/홈에서 쓰는 경로 조각. 작성자 아이디가 아니다.
_NAVER_NON_AUTHOR = {
    "postview.naver", "postview.nhn", "postlist.naver", "postlist.nhn",
    "prologue", "guestbook", "postsearchlist.naver",
}


class BlogRef(NamedTuple):
    """주소에서 읽어낸 것.

    platform: 'naver' | 'tistory' | 'brunch' | '' (해석 불가)
    author:   블로그 주인 식별자 (소문자). 못 읽으면 ''
    post_id:  글 식별자. 홈/목록 주소면 None
    """
    platform: str
    author: str
    post_id: Optional[str]

    @property
    def is_post(self) -> bool:
        """개별 글을 가리키는 주소인가. 홈 주소면 False."""
        return bool(self.platform) and bool(self.post_id)


def parse_blog_ref(url) -> BlogRef:
    """블로그 주소 → (플랫폼, 작성자, 글 번호). 해석 못 하면 빈 값."""
    if not isinstance(url, str) or not url.strip():
        return BlogRef("", "", None)

    parsed = urlparse(url.strip())
    if parsed.scheme not in ("http", "https"):
        return BlogRef("", "", None)

    host = (parsed.hostname or "").lower()
    parts = [p for p in parsed.path.split("/") if p]
    query = parse_qs(parsed.query or "")

    def q(name):
        v = query.get(name) or []
        return v[0].strip() if v and v[0].strip() else None

    # 네이버 — 호스트는 **정확히** 일치해야 한다 (부분 문자열 금지)
    if host in NAVER_HOSTS:
        author = q("blogId") or ""
        post_id = q("logNo")
        if not author and parts and parts[0].lower() not in _NAVER_NON_AUTHOR:
            author = parts[0]
            if len(parts) >= 2 and parts[1].isdigit():
                post_id = parts[1]
        return BlogRef("naver", author.lower(), post_id)

    # 티스토리 — {author}.tistory.com/{postNo}
    if host == "tistory.com" or host.endswith(".tistory.com"):
        author = host[: -len(".tistory.com")] if host.endswith(".tistory.com") else ""
        post_id = None
        if parts:
            # /123 또는 /entry/제목 둘 다 개별 글이다
            post_id = parts[0] if parts[0].isdigit() else ("/".join(parts) if parts[0] != "category" else None)
        return BlogRef("tistory", author.lower(), post_id)

    # 브런치 — brunch.co.kr/@{author}/{no}
    if host in BRUNCH_HOSTS:
        author = ""
        post_id = None
        if parts:
            first = parts[0]
            author = first[1:] if first.startswith("@") else first
            if len(parts) >= 2:
                post_id = parts[1]
        return BlogRef("brunch", author.lower(), post_id)

    return BlogRef("", "", None)


def same_blog(registered_url, requested_url) -> bool:
    """등록된 블로그와 요청된 주소가 같은 사람의 블로그인가.

    플랫폼과 작성자가 모두 일치해야 한다. 작성자를 못 읽으면 통과시키지 않는다 —
    빈 값끼리 같다고 판단하면 아무 주소나 통과한다.
    """
    a = parse_blog_ref(registered_url)
    b = parse_blog_ref(requested_url)
    if not a.platform or not a.author:
        return False
    return a.platform == b.platform and a.author == b.author
